### ISSUE TITLE
Debian bug "I require modinfo but it's not installed", usr-merged system

### ISSUE THREAD

**Issue by markinfo** [graysky2/profile-sync-daemon#258]
Hi.
There is [bug](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=932345) in Debian with missing modinfo.
As Shengjing Zhu already written, there should be (at least for Debian) in /usr/bin/profile-sync-daemon
`
[[ -h /sbin ]] || PATH=$PATH:/sbin
`
changed to
`
[[ -h /usr/sbin ]] || PATH=$PATH:/usr/sbin
`
Should that be in upstream corrected?

**Comment by graysky2** [comment]
So if I understand correctly, you're recommending removing this line:
```
[[ -h /sbin ]] || PATH=$PATH:/sbin
```

And replacing it with this line:
```
[[ -h /usr/sbin ]] || PATH=$PATH:/usr/sbin
```

**Comment by markinfo** [comment]
profile-sync-daemon does not depend on kmod. But linux-image depends on kmod.
The problem with missing modinfo comes after
`psd p` command.

**Comment by graysky2** [comment]
Regardless of the PATH within the script, if psd is installed on a debian system, `modinfo` needs to be present as well.  I spun up a buster VM (lxc) and it is not pulled in by whatever the base set of packages is nor is it pulled in by apt-get upon installation of psd so that is more serious to me.

As an aside, I will move the dependency checks up in the script so it will fail before any options are parsed.

```
$ ssh test@deb
test@10.0.3.96's password:
Linux deb 5.6.5-1-minimum #1 SMP PREEMPT Sun, 19 Apr 2020 14:40:15 +0000 x86_64

The programs included with the Debian GNU/Linux system are free software;
the exact distribution terms for each program are described in the
individual files in /usr/share/doc/*/copyright.

Debian GNU/Linux comes with ABSOLUTELY NO WARRANTY, to the extent
permitted by applicable law.
Last login: Tue Apr 21 12:06:19 2020 from 10.0.3.1
$ su -
Password: 
root@deb:~# uname -a
Linux deb 5.6.5-1-minimum #1 SMP PREEMPT Sun, 19 Apr 2020 14:40:15 +0000 x86_64 GNU/Linux

root@deb:~# apt-get install profile-sync-daemon
Reading package lists... Done
Building dependency tree       
Reading state information... Done
The following additional packages will be installed:
  libpopt0 rsync
The following NEW packages will be installed:
  libpopt0 profile-sync-daemon rsync
0 upgraded, 3 newly installed, 0 to remove and 0 not upgraded.
Need to get 468 kB of archives.
After this operation, 1,067 kB of additional disk space will be used.
Do you want to continue? [Y/n] 
Get:1 http://deb.debian.org/debian buster/main amd64 libpopt0 amd64 1.16-12 [49.4 kB]
Get:2 http://deb.debian.org/debian buster/main amd64 rsync amd64 3.1.3-6 [397 kB]
Get:3 http://deb.debian.org/debian buster/main amd64 profile-sync-daemon all 6.31-1 [21.4 kB]
Fetched 468 kB in 0s (2,100 kB/s)              
debconf: delaying package configuration, since apt-utils is not installed
Selecting previously unselected package libpopt0:amd64.
(Reading database ... 13889 files and directories currently installed.)
Preparing to unpack .../libpopt0_1.16-12_amd64.deb ...
Unpacking libpopt0:amd64 (1.16-12) ...
Selecting previously unselected package rsync.
Preparing to unpack .../rsync_3.1.3-6_amd64.deb ...
Unpacking rsync (3.1.3-6) ...
Selecting previously unselected package profile-sync-daemon.
Preparing to unpack .../profile-sync-daemon_6.31-1_all.deb ...
Unpacking profile-sync-daemon (6.31-1) ...
Setting up libpopt0:amd64 (1.16-12) ...
Setting up rsync (3.1.3-6) ...
Created symlink /etc/systemd/system/multi-user.target.wants/rsync.service → /lib/systemd/system/rsync.service.
Setting up profile-sync-daemon (6.31-1) ...
Processing triggers for systemd (241-7~deb10u3) ...
Processing triggers for libc-bin (2.28-10) ...
root@deb:~# exit
logout

$ psd
 First time running psd so please edit /home/test/.config/psd/psd.conf to your liking and run again.

$ psd 
Profile-sync-daemon v6.31 on Debian GNU/Linux 10 (buster)

 /usr/bin/psd [option]
  preview	Parse config file (/home/tit/.config/psd/psd.conf) to see which profiles will be managed.
  clean		Clean (delete without prompting) ALL crashrecovery dirs for all profiles.

 It is HIGHLY DISCOURAGED to directly call /usr/bin/psd to sync, resync, or to unsync.

 Instead, use systemd to start/stop profile-sync-daemon.

 systemctl --user [option] psd
  start		Turn on daemon; make symlinks and actively manage targets in tmpfs.
  stop		Turn off daemon; remove symlinks and rotate tmpfs data back to disc.
  enable	Autostart daemon when system comes up.
  disable	Remove daemon from the list of autostart daemons.

$ psd p
 I require modinfo but it's not installed. Aborting!

$ modinfo
-sh: 5: modinfo: not found
```

**Comment by sr500** [comment]
Fresh install debian 10.3, kmod was automaticly installed. Same error:

$ psd p
 I require modinfo but it's not installed. Aborting!

#apt policy kmod
kmod:
  Installed: 26-1
  Candidate: 26-1
  Version table:
 *** 26-1 500
        500 http://deb.debian.org/debian buster/main amd64 Packages
        100 /var/lib/dpkg/status

uname -a
Linux 4.19.0-8-amd64 #1 SMP Debian 4.19.98-1 (2020-01-26) x86_64

**Comment by sr500** [comment]
I`ve found the workaraound on https://github.com/graysky2/profile-sync-daemon/issues/253

https://github.com/graysky2/profile-sync-daemon/issues/253#issuecomment-589663837

It works!

btw: in my /usr/bin/profile-sync-daemon is it in line 190

**Comment by graysky2** [comment]
@sr500 - Are you saying that you modified your psd to include 2 checks?  The current version does this natively: https://github.com/graysky2/profile-sync-daemon/blob/9cdc37c0f540d7c1c70d3998eb8ca0445a49505e/common/profile-sync-daemon.in#L203

Debian is shipping an older version where this isn't the case.  If you're saying that is the fix for you, I can't force them to bump their version.

As a sanity check, can you build the latest git version and verify that it works for you?

**Comment by sr500** [comment]
**@graysky2** 

No, i have changed in Version:

#apt policy profile-sync-daemon
profile-sync-daemon:
  Installed: 6.31-1
  Candidate: 6.31-1
  Version table:
 *** 6.31-1 500
        500 http://deb.debian.org/debian buster/main amd64 Packages
        100 /var/lib/dpkg/status

/usr/bin/profile-sync-daemon
190 command -v  modinfo >/dev/null 2>&1 || {
to
190 command -v  /sbin/modinfo >/dev/null 2>&1 || {

On another system is running Debian bullseye sid:

#apt policy kmod
kmod:
  Installed: 27+20200310-2
  Candidate: 27+20200310-2
  Version table:
 *** 27+20200310-2 500
        500 http://deb.debian.org/debian unstable/main amd64 Packages
        100 /var/lib/dpkg/status

#apt policy profile-sync-daemon
profile-sync-daemon:
  Installed: 6.34-1
  Candidate: 6.34-1
  Version table:
 *** 6.34-1 500
        500 http://deb.debian.org/debian unstable/main amd64 Packages
        100 /var/lib/dpkg/status

Here does psd works without problems: 

$psd p
Profile-sync-daemon v6.34 on Debian GNU/Linux bullseye/sid

/usr/bin/profile-sync-daemon
190   command -v modinfo >/dev/null 2>&1 || {

**Comment by graysky2** [comment]
So we can close this?

**Comment by sr500** [comment]
Yes for me it's ok

Thx for helping!

**Comment by zhsj** [comment]
Hi @graysky2 

The fix is not correct if I read the code correctly.

The script https://github.com/graysky2/profile-sync-daemon/blob/master/common/profile-sync-daemon.in calls directly `modinfo` in https://github.com/graysky2/profile-sync-daemon/blob/6d1cac08176b79696eead82707617ed86220342b/common/profile-sync-daemon.in#L173 and other places. So `modinfo` needs to be in `PATH`.

But the default `PATH` for Debian is `/usr/local/bin:/usr/bin:/bin:/usr/local/games:/usr/games`, which doesn't have `/sbin` or `/usr/sbin`. But `modinfo` is in `/sbin/modinfo`.

So psd try to add `sbin` to `PATH` in https://github.com/graysky2/profile-sync-daemon/blob/6d1cac08176b79696eead82707617ed86220342b/common/profile-sync-daemon.in#L11

However this doesn't work when `/sbin` is a symlink to `/usr/sbin`(usr-merged system). Because the first condition `-h /sbin` is true, so it doesn't add `/sbin` to `PATH`.

The easied way I think is to remove the test, just add `/sbin` to `PATH` regardless. There's no harm if you have duplicated `/sbin` in `PATH`.

**Comment by zhsj** [comment]
The fix in https://github.com/graysky2/profile-sync-daemon/commit/117126530161438bdd04ddeeab47405bbec46f6d just hides the error.

**Comment by graysky2** [comment]
I think you're correct, thank you for the detailed explanation.

**Comment by Mr-Ruben** [comment]
Correct me if I am wrong, but ...

The message in https://www.mail-archive.com/debian-bugs-dist@lists.debian.org/msg1688752.html

says

> In in debian modinfo is located:
> [/usr]/sbin/modinfo
> 
> Workaround: change 'modinfo ...' to '/sbin/modinfo ...' in 
> /usr/bin/profile-sync-daemon


Which is **NOT** just this line like I have read  in 
https://github.com/graysky2/profile-sync-daemon/issues/253#issuecomment-589663837
or
https://github.com/graysky2/profile-sync-daemon/issues/258#issuecomment-619194054

```
# No, this is not enough
command -v /sbin/modinfo >/dev/null 2>&1 || {

```

As zhsj very well said in https://github.com/graysky2/profile-sync-daemon/issues/258#issuecomment-633264877

> The fix in 1171265 just hides the error.

https://github.com/graysky2/profile-sync-daemon/commit/117126530161438bdd04ddeeab47405bbec46f6d

(I believe) The fix is to edit ALL references to "modinfo" and convert them to "/sbin/modinfo" , which is what the first link said very briefly.

In my file there are 3 + 1 references to edit (already changed below)

```
$ grep modinfo  /usr/bin/profile-sync-daemon 
# module is in the tree using modinfo
  /sbin/modinfo overlayfs &>/dev/null
  /sbin/modinfo overlay &>/dev/null
  command -v /sbin/modinfo >/dev/null 2>&1 || {
  echo -e " ${BLD}I require modinfo but it's not installed. ${RED}Aborting!${NRM}" >&2; exit 1; }

```
